curvefs/src/metaserver/s3infocache.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * @Project: curve
+ * @Date: 2021-11-02
+ * @Author: majie1
+ */
+
+#include "curvefs/src/metaserver/s3infocache.h"
+
+#include "curvefs/proto/mds.pb.h"
+
+namespace curvefs {
+namespace metaserver {
+
+void S3InfoCache::UpdateRecent(uint64_t fsid) {
+    if (pos_.find(fsid) != pos_.end()) {
+        recent_.erase(pos_[fsid]);
+    } else if (recent_.size() >= capacity_) {
+        auto old = recent_.back();
+        recent_.pop_back();
+        cache_.erase(old);
+        pos_.erase(old);
+    }
+    recent_.push_front(fsid);
+    pos_[fsid] = recent_.begin();
+}
+
+S3Info S3InfoCache::Get(uint64_t fsid) {
+    UpdateRecent(fsid);
+    return cache_[fsid];
+}
+
+void S3InfoCache::Put(uint64_t fsid, const S3Info& s3info) {
+    UpdateRecent(fsid);
+    cache_[fsid] = s3info;
+}
+
+S3InfoCache::RequestStatusCode S3InfoCache::RequestS3Info(uint64_t fsid,
+                                                          S3Info* s3info) {
+    // send GetFsInfoRequest to mds
+    curvefs::mds::GetFsInfoRequest request;
+    curvefs::mds::GetFsInfoResponse response;
+    request.set_fsid(fsid);
+    brpc::Channel channel;
+    if (channel.Init(mdsAddrs_[mdsIndex_].c_str(), NULL) != 0) {
```

Comment:
do not do any retry here?

Reply:
I think there is no necessity maybe. just let it fail.
retry S3InfoCache::GetS3Info high-level rather than retry in S3InfoCache maybe better.